baroclinic-adjustment/baroclinic_adjustment_reactant.jl

```diff
@@ -0,0 +1,164 @@
+using Oceananigans
+using Reactant
+using Oceananigans.Units
+using Oceananigans.Architectures: ReactantState
+using Random
+using Printf
+using JSON
+
+# Architecture
+# use_reactant = get(ENV, "use-reactant", false)
+# if use_reactant == "true" || use_reactant=="1"
+#     use_reactant = true
+# elseif use_reactant == "false" || use_reactant=="0"
+#     use_reactant = false
+# end
+
+# raise = get(ENV, "raise", false)
+# if raise == "true" || raise=="1"
+#     raise = true
+# elseif raise == "false" || raise=="0"
+#     raise = false
+# end
+
+# if use_reactant
+#     arch = Oceananigans.Architectures.ReactantState()
+# else
+#     arch = GPU() # change this to use CPU
+# end
+# if raise
+    Reactant.Compiler.Raise[] = true
+# end
+
+function loop!(model, Δt, Ninner)
+    # Δt = 1200 # 20 minutes
+    Oceananigans.TimeSteppers.first_time_step!(model, Δt)
+    @trace for _ = 2:Ninner
```

Comment:
```suggestion
    @trace for _ = 1:Ninner
```

if making this change, then we do have to be sure to call `first_time_step!` at the very beginning before initiating the main loop

Reply:
I am confused now. Isn't the "main loop" over `Ninner` ?

Reply:
No, because we cannot print / log OR emit output from an `@compile`'d function.

Reply:
Here is a simple implementation of an outer loop, as an example:

```julia
Δt = 60 # seconds
Ninner = ConcreteRNumber(10) # this must be traced, so that we can change it later.
r_loop! = @compile loop!(model, Δt, Ninner)
r_first_time_step! = @compile Oceananigans.TimeSteppers.first_time_step!(model, Δt)

# Do one initial loop
Nprint = 10
r_first_time_step!(model, Δt)
r_loop!(model, Δt, ConcreteRNumber(Nprint-1))

# Now set an outer loop in motion
Ntotal = 1000 # or, compute this with a stop_time + Δt
Nouter = ceil(Int, (Ntotal - 1) / Nprint)

for outer = 1:Nouter
    r_loop!(model, Δt, ConcreteRNumber(Nprint))
    print_something_useful(model)
end
```

Reply:
Thanks - got the construct 😅

Reply:
Basically the idea is we are leaving Simulation behind and sailing into a brave new world of manual loops!! Interestingly, this is how Oceananigans was used during very early days. We introduced Simulation after a few years.

Reply:
Note it also req changing the `progress` functions to act on `model` not `sim`. Hope this is easy because I believe they only req info from `sim.model`. The output writers also act on `model` via `Oceananigans.OutputWriters.write_output!(output_writer, model)`

Reply:
I see. For now I tried to implement "manual" progress without using callbacks as I was getting an error their not being supported by Reactant.